libevmasm/ThirdOptimizerTesting.cpp

```diff
@@ -0,0 +1,313 @@
+			}
+			else
+				assertThrow(false, OptimizerException, "cannot create items from non-bound operation pattern without operation values");
+		case NewOptimizerPattern::Kind::Unknown:
+			assertThrow(false, OptimizerException, "cannot create items from non-bound unknown pattern");
+		default:
+			assertThrow(false, OptimizerException, "unreachable");
+	}
+}
+
+
+bool dev::solidity::NewOptimizerPattern::matches(std::vector<AssemblyItem> const& _items, bool _unbind)
+{
+	if (_unbind) unbind();
+
+	if (isBound())
+		return boundItems() == _items;
+
+	switch (kind())
+	{
+		case Kind::Constant:
+		{
+			if (_items.size() != 1) return false;
+
+			if (_items.back().type() != AssemblyItemType::Push) return false;
+
+			if (hasConstant())
+			{
+				if (_items[0].data() == constant())
+				{
+					bind(_items);
+					return true;
+				}
+
+				return false;
+			}
+			else
+			{
+				bind(_items);
+				return true;
+			}
+		}
+		case Kind::Operation:
+		{
+			if (!hasOperationValues())
+			{
+				if (_items.back().type() == AssemblyItemType::Operation)
+				{
+					bind(_items);
+					return true;
+				}
+
+				return false;
+			}
+
+			auto const instruction = this->instruction();
+
+			if (_items.back().type() != AssemblyItemType::Operation) return false;
+			if (_items.back().instruction() != instruction) return false;
+
+			auto const argumentCount = instructionInfo(instruction).args;
+			assertThrow(unsigned(argumentCount) == operands().size(), OptimizerException, "");
+
+			if (argumentCount == 0) return _items.size() == 1;
+
+			auto parsedArguments = parseArguments(_items);
+
+			auto& requiredArgumentPatterns = operands();
+
+			assertThrow(
+				requiredArgumentPatterns.size() == parsedArguments.size(),
+				OptimizerException,
+				"required argument count and parsed argument count did not match"
+			);
+
+			for (unsigned i = 0; i < parsedArguments.size(); i++)
+			{
+				auto& argumentItems = parsedArguments[i];
+
+				if (!requiredArgumentPatterns[i].matches(argumentItems, false)) return false;
+			}
+
+			bind(_items);
+
+			return true;
+		}
+		case Kind::Any:
+			bind(_items);
+			return true;
+		case Kind::Unknown:
+			return false;
+		default:
+			assertThrow(false, OptimizerException, "invalid pattern kind");
+	}
+}
+
+void NewOptimizerPattern::bind(std::vector<AssemblyItem> const& _items)
+{
+	if (kind() == Kind::Constant)
+	{
+		assertThrow(_items.size() == 1 && _items.back().type() == AssemblyItemType::Push, OptimizerException, "invalid bind type");
+		if (hasConstant())
+			assertThrow(constant() == _items.back().data(), OptimizerException, "invalid bind constant");
+	}
+
+	if (kind() == Kind::Operation)
+	{
+		assertThrow(!_items.empty() && _items.back().type() == AssemblyItemType::Operation, OptimizerException, "invalid bind value");
+		if (hasOperationValues())
+			assertThrow(instruction() == _items.back().instruction(), OptimizerException, "invalid bind instruction");
+	}
+
+	m_ptr->m_boundItems = _items;
```

Comment:
This should not copy, if possible.

Reply:
Well, it's passed by const ref, so a copy is necessary here. The other option would be to pass by value and move into place.

Reply:
`m_boundItems` is not a reference, so it will copy.

Reply:
I meant `_items`. I'm not sure `m_boundItems` can be a reference, that seems like it might cause lifetime issues, but I could be totally wrong.

Reply:
Indeed, it cannot be a reference, that's why I would propose to work with offsets into the existing item vector.

Reply:
If we were to use offsets there, we would have to use offsets/iterators basically everywhere, since right now there is not one singular items vector, but a bunch of vectors copied from each other.